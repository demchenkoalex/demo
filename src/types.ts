export enum FetchingState {
  Loading,
  LoadingMore,
  None,
  Refreshing
}

export interface PaginatedResponse<T> {
  data: T[]
  page: number
  per_page: number
  total: number
  total_pages: number
}

export interface User {
  avatar: string
  first_name: string
  id: number
  last_name: string
}
