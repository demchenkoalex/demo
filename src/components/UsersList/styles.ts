import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  loadingContainer: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center'
  }
})
