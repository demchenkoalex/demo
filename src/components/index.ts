export * from './UserRow'
export * from './UsersList'
