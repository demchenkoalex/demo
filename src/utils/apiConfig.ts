import axios from 'axios'

const api = axios.create({
  baseURL: 'https://reqres.in/api',
  timeout: 10000
})

export default api
