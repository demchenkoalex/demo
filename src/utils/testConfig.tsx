import React, { ComponentType } from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

export const setupJestShallow = ({ Component, props }: { Component: ComponentType<any>, props: any }) => {
  const shallowRenderer = createRenderer()

  shallowRenderer.render(<Component {...props} />)

  const wrapper = shallowRenderer.getRenderOutput()
  const instance = shallowRenderer.getMountedInstance() as any

  return {
    wrapper: { ...wrapper, instance: () => instance }
  }
}
