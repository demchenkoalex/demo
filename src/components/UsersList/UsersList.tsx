import React, { PureComponent } from 'react'
import {
  ActivityIndicator,
  Alert,
  FlatList,
  Text,
  View
} from 'react-native'
import {
  FetchingState,
  PaginatedResponse,
  User
} from 'src/types'
import api from 'src/utils/apiConfig'
import { UserRow } from '../UserRow'
import styles from './styles'

export interface Props {
  itemsPerPage: number
  onPageChange: (page: number) => void
}

interface State {
  fetchingState: FetchingState,
  page: number,
  totalPages: number,
  users: User[]
}

export class UsersList extends PureComponent<Props, State> {
  state = {
    fetchingState: FetchingState.None,
    page: 1,
    totalPages: 1,
    users: []
  }

  componentDidMount = () => this.load()

  fetchData = async (fetchingState: FetchingState) => {
    const { itemsPerPage, onPageChange } = this.props
    const { page, users } = this.state
    const resetData =
      fetchingState === FetchingState.Loading ||
      fetchingState === FetchingState.Refreshing

    try {
      this.setState({ fetchingState })

      const { data } = await api.get<PaginatedResponse<User>>('/users', {
        params: {
          page: resetData ? 1 : page + 1,
          per_page: itemsPerPage
        }
      })

      this.setState({
        page: data.page,
        totalPages: data.total_pages,
        users: resetData ? data.data : [...users, ...data.data]
      }, this.handlePageChange)
    } catch (error) {
      Alert.alert('Error', error.message, [{ text: 'OK' }], { cancelable: true })
    } finally {
      this.setState({ fetchingState: FetchingState.None })
    }
  }

  handlePageChange = () => this.props.onPageChange(this.state.page)

  keyExtractor = (item: User) => item.id.toString()

  load = async () => {
    const { fetchingState, users } = this.state

    if (fetchingState !== FetchingState.None || users.length > 0) return

    await this.fetchData(FetchingState.Loading)
  }

  loadMore = async () => {
    const { fetchingState, page, totalPages, users } = this.state

    if (page >= totalPages || fetchingState !== FetchingState.None || users.length === 0) return

    await this.fetchData(FetchingState.LoadingMore)
  }

  refresh = async () => {
    const { fetchingState } = this.state

    if (fetchingState !== FetchingState.None) return

    await this.fetchData(FetchingState.Refreshing)
  }

  renderFooter = () => this.state.fetchingState === FetchingState.LoadingMore
    ? <ActivityIndicator style={{ paddingVertical: 16 }} />
    : <View style={{ height: 32 }} />

  renderItem = ({ item }: { item: User }) => <UserRow user={item} />

  render () {
    const { fetchingState, users } = this.state

    return (
      fetchingState === FetchingState.Loading
      ?
        <View style={styles.loadingContainer}>
          <ActivityIndicator/>
        </View>
      :
        <FlatList
          data={users}
          extraData={fetchingState}
          keyExtractor={this.keyExtractor}
          ListFooterComponent={this.renderFooter}
          onEndReached={this.loadMore}
          onEndReachedThreshold={0.2}
          onRefresh={this.refresh}
          refreshing={fetchingState === FetchingState.Refreshing}
          renderItem={this.renderItem}
        />
    )
  }
}
