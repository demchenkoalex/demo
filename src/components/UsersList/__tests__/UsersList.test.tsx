/* tslint:disable */

import MockAdapter from 'axios-mock-adapter'
import React from 'react'
import { ActivityIndicator, View } from 'react-native'
import { UserRow } from 'src/components'
import { FetchingState } from 'src/types'
import api from 'src/utils/apiConfig'
import { setupJestShallow } from 'src/utils/testConfig'
import { Props, UsersList } from '../UsersList'

describe('UsersList', () => {
  const props: Props = {
    itemsPerPage: 5,
    onPageChange: jest.fn()
  }

  describe('Shapshots', () => {
    it('with default props', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Methods', () => {
    it('has componentDidMount defined', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      expect(typeof wrapper.instance().componentDidMount).toBe('function')
      expect(wrapper.instance().componentDidMount()).resolves.toBe(undefined)
    })

    it('has fetchData defined', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      const paginatedResponse = { page: 1, total_pages: 3, data: [{ id: 1 }] }
      const mock = new MockAdapter(api)
      mock.onGet('/users').reply(200, paginatedResponse)

      expect(typeof wrapper.instance().fetchData).toBe('function')
      expect(wrapper.instance().fetchData(FetchingState.None)).resolves.toBe(undefined)
    })

    it('has fetchData defined for fetching state loading', async () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      const paginatedResponse = { page: 1, total_pages: 3, data: [{ id: 1 }] }
      const mock = new MockAdapter(api)
      mock.onGet('/users').reply(200, paginatedResponse)

      expect(typeof wrapper.instance().fetchData).toBe('function')
      expect(wrapper.instance().fetchData(FetchingState.Loading)).resolves.toBe(undefined)
    })

    it('has fetchData defined for unsuccessful response', async () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      const mock = new MockAdapter(api)
      mock.onGet('/users').reply(404)

      expect(typeof wrapper.instance().fetchData).toBe('function')

      try {
        await wrapper.instance().fetchData(FetchingState.None)
      } catch (error) {
        expect(error).toBeDefined()
      }
    })

    it('has handlePageChange defined', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      expect(typeof wrapper.instance().handlePageChange).toBe('function')
      expect(wrapper.instance().handlePageChange()).toBe(undefined)
    })

    it('has keyExtractor defined', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      expect(typeof wrapper.instance().keyExtractor).toBe('function')
      expect(wrapper.instance().keyExtractor({ id: 1 })).toBe('1')
    })

    it('has load defined', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      expect(typeof wrapper.instance().load).toBe('function')
      expect(wrapper.instance().load()).resolves.toBe(undefined)
    })

    it('has load defined for fetching state loading', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      wrapper.instance().setState({ fetchingState: FetchingState.Loading })
      expect(typeof wrapper.instance().load).toBe('function')
      expect(wrapper.instance().load()).resolves.toBe(undefined)
    })

    it('has load defined for non empty users', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      wrapper.instance().setState({ users: [{ id: 1 }] })
      expect(typeof wrapper.instance().load).toBe('function')
      expect(wrapper.instance().load()).resolves.toBe(undefined)
    })

    it('has loadMore defined', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      wrapper.instance().setState({ page: 1, totalPages: 2, fetchingState: FetchingState.None, users: [{ id: 1 }] })
      expect(typeof wrapper.instance().loadMore).toBe('function')
      expect(wrapper.instance().loadMore()).resolves.toBe(undefined)
    })

    it('has loadMore defined for last page', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      wrapper.instance().setState({ page: 2, totalPages: 2 })
      expect(typeof wrapper.instance().loadMore).toBe('function')
      expect(wrapper.instance().loadMore()).resolves.toBe(undefined)
    })

    it('has loadMore defined for fetching state loading more', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      wrapper.instance().setState({ fetchingState: FetchingState.LoadingMore })
      expect(typeof wrapper.instance().loadMore).toBe('function')
      expect(wrapper.instance().loadMore()).resolves.toBe(undefined)
    })

    it('has loadMore defined for empty users', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      wrapper.instance().setState({ users: [] })
      expect(typeof wrapper.instance().loadMore).toBe('function')
      expect(wrapper.instance().loadMore()).resolves.toBe(undefined)
    })

    it('has refresh defined', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      expect(typeof wrapper.instance().refresh).toBe('function')
      expect(wrapper.instance().refresh()).resolves.toBe(undefined)
    })

    it('has refresh defined for fetching state refreshing', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      wrapper.instance().setState({ fetchingState: FetchingState.Refreshing })
      expect(typeof wrapper.instance().refresh).toBe('function')
      expect(wrapper.instance().refresh()).resolves.toBe(undefined)
    })

    it('has renderFooter defined', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      expect(typeof wrapper.instance().renderFooter).toBe('function')
      expect(wrapper.instance().renderFooter()).toEqual(<View style={{ height: 32 }} />)
    })

    it('has renderFooter defined for loading more state', () => {
      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      wrapper.instance().setState({ fetchingState: FetchingState.LoadingMore })
      expect(typeof wrapper.instance().renderFooter).toBe('function')
      expect(wrapper.instance().renderFooter()).toEqual(<ActivityIndicator style={{ paddingVertical: 16 }} />)
    })

    it('has renderItem defined', () => {
      const user = {
        avatar: 'https://avatar.com/alex',
        first_name: 'Alex',
        id: 1,
        last_name: 'Demchenko'
      }

      const { wrapper } = setupJestShallow({
        Component: UsersList,
        props
      })

      expect(typeof wrapper.instance().renderItem).toBe('function')
      expect(wrapper.instance().renderItem({ item: user })).toEqual(<UserRow user={user} />)
    })
  })
})

/* tslint:enable */
