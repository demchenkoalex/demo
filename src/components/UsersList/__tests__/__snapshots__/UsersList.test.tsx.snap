// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList Shapshots with default props 1`] = `
<FlatList
  ListFooterComponent={[Function]}
  data={Array []}
  disableVirtualization={false}
  extraData={2}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReached={[Function]}
  onEndReachedThreshold={0.2}
  onRefresh={[Function]}
  refreshing={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
